### PR TITLE
feat(managers/npm): add entries with protocol prefix in temporary .yarnrc.yml file

### PR DESCRIPTION
**Pull request by secustor**
## Changes
Add entries with protocol prefix, it supplied to temporary`.yarnrc.yml`. 
I have done also some small refactoring. 

## Context
Yarn allows to pin to add the protocol prefix to the host, though does merge configuration if only the host is supplied. 
This is similar to our host rules logic. 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/npm/post-update/rules.ts [hunk 1]
```diff
@@ -3,51 +3,83 @@ import * as hostRules from '../../../../util/host-rules';
...
+    const matchedHost = hostRule.matchHost;
+    // Should never be needed as if we have a resolvedHost there has to be a matchHost
+    // istanbul ignore next
+    if (!matchedHost) {
+      continue;
+    }
+
+    const uri = `//${matchedHost}/`;
+    let cleanedUri = uri;
+    if (is.string(matchedHost) && isHttpUrl(matchedHost)) {
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/rules.ts:35]: why the additional is.string check?

File: lib/modules/manager/npm/post-update/rules.ts [hunk 2]
```diff
@@ -3,51 +3,83 @@ import * as hostRules from '../../../../util/host-rules';
...
+      const registries = {
+        npmAuthIdent: `${hostRule.username}:${hostRule.password}`,
+      };
+      additionalYarnRcYml.npmRegistries[cleanedUri] = registries;
+      additionalYarnRcYml.npmRegistries[uri] = registries;
     }
   }
-  return { additionalNpmrcContent, additionalYarnRcYml };
+
+  const hasYarnRCNPMRegistries =
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/rules.ts:77]:
```suggestion
  const hasYarnRcNpmRegistries =
```
be consistent with interface

File: lib/modules/manager/npm/post-update/rules.ts [hunk 3]
```diff
@@ -3,51 +3,83 @@ import * as hostRules from '../../../../util/host-rules';
...
+    Object.keys(additionalYarnRcYml.npmRegistries).length > 0;
+  return {
+    additionalNpmrcContent,
+    additionalYarnRcYml: hasYarnRCNPMRegistries
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/rules.ts:81]:
```suggestion
    additionalYarnRcYml: hasYarnRcNpmRegistries
```

File: lib/modules/manager/npm/post-update/rules.ts [hunk 4]
```diff
@@ -3,51 +3,83 @@ import * as hostRules from '../../../../util/host-rules';
...
+      const registry = {
+        npmAuthToken: hostRule.token,
+      };
+      additionalYarnRcYml.npmRegistries[cleanedUri] = registry;
+      additionalYarnRcYml.npmRegistries[uri] = registry;
+
+      continue;
+    }
+
+    if (is.string(hostRule.username) && is.string(hostRule.password)) {
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/rules.ts:62]: use else if, so we don't need continue above? less code.

Comment (secustor) [reply lib/modules/manager/npm/post-update/rules.ts:62]: That's hurting readability as we get a lot of indentations wrapping here.

File: lib/modules/manager/npm/post-update/rules.ts (hunk 1 above)
Comment (secustor) [reply lib/modules/manager/npm/post-update/rules.ts]: left over from before the falsely check has been added.

**Review by rarkins — APPROVED** [review]

**Pull request merged**